### PR TITLE
Transform the test-commons API in two main interfaces SingleFileVerifier and MultiFileVerifier

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NoncompliantCommentParser.java#L40](https://github.com/SonarSource/sonar-analyzer-commons/blob/8925746f8a607400e22e43a6b3b7ddcf0e092c03/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/NoncompliantCommentParser.java#L40): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [PreciseLocationParser.java#L41](https://github.com/SonarSource/sonar-analyzer-commons/blob/8925746f8a607400e22e43a6b3b7ddcf0e092c03/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/PreciseLocationParser.java#L41): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * Add a flow location with an optional message
+     * @param line, start at 1, line number of the first character
+     * @param column, start at 1, column number of the first character
+     * @param endLine, start at 1, line number of the last character, if the location is on one line then endLine == line
+     * @param endColumn, start at 1, column number of the last character, if there's only one character then endColumn == column
+     * @param flowIndex, start at 1, to distinguish on which flow the location is appended
+     * @param message optional message, can be null
+     */
+    Issue addFlow(int line, int column, int endLine, int endColumn, int flowIndex, @Nullable String message);
```
Comment (pynicolas) [inline test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java]: Does it work? If not, we should not expose this method.

File: test-commons/src/test/java/com/sonarsource/checks/verifier/SingleFileVerifierTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package com.sonarsource.checks.verifier;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.junit.Test;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class SingleFileVerifierTest {
```
Comment (pynicolas) [inline test-commons/src/test/java/com/sonarsource/checks/verifier/SingleFileVerifierTest.java]: Do we assert anything in this test?

File: test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ *   verifier.assertNoIssues();
+ * </pre>
+ */
+public interface SingleFileVerifier {
+
+  /**
+   * @param mainSourceFilePath
+   * @param encoding encoding used to load source file
+   */
+  static SingleFileVerifier create(Path mainSourceFilePath, Charset encoding) {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java]: No good reason. In SingleFileVerifier there's only one file, so I'm renaming the param to "sourceFilePath".

File: test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java (hunk 1 above)
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/SingleFileVerifier.java]: Removed.

File: test-commons/src/main/java/com/sonarsource/checks/verifier/internal/TestFile.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+class TestFile extends FileContent {
+
+  private final Map<Integer, List<Comment>> commentListPreLineMap = new HashMap<>();
+
+  TestFile(FileContent file) {
+    super(file.getPath(), file.getContent());
+  }
+
+  void addCommentToHide(Comment comment) {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/internal/TestFile.java]: This method is used to add comment locations of the "Noncompliant" comments that we want to remove from the report using the bellow method "lineWithoutComment(line)".
I'm renaming this method "addNoncompliantCommentLocation"

File: test-commons/src/main/java/com/sonarsource/checks/verifier/internal/Report.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    context = new StringBuilder();
+    expected = new StringBuilder();
+    actual = new StringBuilder();
+  }
+
+  public int getExpectedCount() {
+    return expectedCount;
+  }
+
+  public void setExpectedCount(int count) {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/internal/Report.java]: Issues. So I'm renaming using "expectedIssueCount" and "actualIssueCount"

File: test-commons/src/test/java/com/sonarsource/checks/verifier/SingleFileVerifierTest.java (hunk 2 above)
Comment (alban-auzeill) [reply test-commons/src/test/java/com/sonarsource/checks/verifier/SingleFileVerifierTest.java]: Yes, it's non regression test to check the interface SingleFileVerifier continue to works.

File: test-commons/src/main/java/com/sonarsource/checks/verifier/internal/FileIssues.java [hunk 6]
```diff
@@ -133,21 +129,26 @@ public Report report() {
       SecondaryLocation orphanSecondary = orphanSecondaryOrFlowLocations.get(0);
       throw new IllegalStateException("Secondary location '>' without next primary location at " + orphanSecondary.range.toString());
     }
+    Report report = new Report();
+
+    report.setExpectedCount(expectedIssueMap.values().stream().mapToInt(issues -> issues.messages.size()).sum());
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/internal/FileIssues.java]: We need it to be able to assertNoIssues() using getExpectedIssueCount().

File: test-commons/src/main/java/com/sonarsource/checks/verifier/internal/InternalIssueVerifier.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+public class InternalIssueVerifier implements MultiFileVerifier, SingleFileVerifier {
+
+  private final Path mainSourceFilePath;
+  private Set<Path> filesToVerify = new LinkedHashSet<>();
+  private Map<Path, List<Comment>> comments = new HashMap<>();
+  private Map<Path, List<InternalIssue>> actualIssues = new HashMap<>();
+  private Charset encoding = StandardCharsets.UTF_8;
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply test-commons/src/main/java/com/sonarsource/checks/verifier/internal/InternalIssueVerifier.java]: Removed.

File: test-commons/src/main/java/com/sonarsource/checks/verifier/internal/InternalIssueVerifier.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package com.sonarsource.checks.verifier.internal;
+
+import com.sonarsource.checks.verifier.FileContent;
+import com.sonarsource.checks.verifier.MultiFileVerifier;
+import com.sonarsource.checks.verifier.SingleFileVerifier;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
```
Comment (SonarTech) [inline test-commons/src/main/java/com/sonarsource/checks/verifier/internal/InternalIssueVerifier.java]: MINOR Remove this unused import 'java.nio.charset.StandardCharsets'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NoncompliantCommentParser.java#L40](https://github.com/SonarSource/sonar-analyzer-commons/blob/b9d4e961070eb8bd24f4eb25fe5df9d7a9ec4770/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/NoncompliantCommentParser.java#L40): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [PreciseLocationParser.java#L41](https://github.com/SonarSource/sonar-analyzer-commons/blob/b9d4e961070eb8bd24f4eb25fe5df9d7a9ec4770/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/PreciseLocationParser.java#L41): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NoncompliantCommentParser.java#L40](https://github.com/SonarSource/sonar-analyzer-commons/blob/722c75a69c03c31acef23ed1dbe959db39501ae3/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/NoncompliantCommentParser.java#L40): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [PreciseLocationParser.java#L41](https://github.com/SonarSource/sonar-analyzer-commons/blob/722c75a69c03c31acef23ed1dbe959db39501ae3/test-commons/src/main/java/com/sonarsource/checks/verifier/internal/PreciseLocationParser.java#L41): This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**